### ISSUE TITLE
Group claims in Azure OAuth token not handled properly if user is a member of > 200 groups

### ISSUE THREAD

**Issue by sivamu** [grafana/grafana#42689]
**What happened**:
I have Grafana setup to use AzureAD OAuth2 authentication, and additionally [configured allowed groups](https://grafana.com/docs/grafana/latest/auth/azuread/#configure-allowed-groups). I tried to login using Azure AD but am getting am getting Access denied as not being a part of the Azure AD group (Error message "User not a member of the required groups") even though I am a member of the security group. 

**What you expected to happen**:
The login should be successful.

**How to reproduce it (as minimally and precisely as possible)**:
1. Enable Azure AD Oauth2
2. Set `allowed_groups` in the Grafana config. This is a security group that you are a part of in AAD
3. Ensure you're a member of more than 200 AAD security groups.
4. Try signing in to Grafana

**Anything else we need to know?**:
I did a bit more research and found that the [`groups` field](https://github.com/grafana/grafana/blob/main/pkg/login/social/azuread_oauth.go#L26) is empty when I am trying to login. I also found that the Azure JWT token doesn't set this `groups` property in the payload if the user is a member of more than 200 AAD groups (just as myself) since it's the limit of group UUIDs that can be included in the token (more info in Azure docs: https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#payload-claims). In this case, `groups` is set to a fixed value, `hasgroups` is set to true and the client would need to query the user's list of entire groups in from the Microsoft Graph API that is provided as a URL in the `groups:src1` property.

Grafana would need to include some logic to query this Graph API URL to get the full list of user's groups and then check for membership, when the `hasgroups` field is set due to the user having too many groups.

**Environment**:
- Grafana version: 8.1.0
- OS Grafana is installed on: Grafana docker image
- User OS & Browser: Mac OS + Edge
- Grafana plugins: N/A
- Others: N/A

**Comment by wbrowne** [comment]
@sivamu Can you share the OAuth config you've used? (and please omit sensitive information)

Can you confirm that the problem doesn't exist when the user is part of < 200 groups?

**Comment by sivamu** [comment]
Sure thing. Here's the config in Grafana:
```
grafana.ini:  |
    [auth]
    disable_login_form = true

    [auth.basic]
    enabled = false

    [auth.azuread]
    name = Azure AD
    enabled = true
    allow_sign_up = true
    client_id = <Redacted Client ID>
    scopes = openid email profile
    auth_url = https://login.microsoftonline.com/<Redacted ID>/oauth2/v2.0/authorize
    token_url = https://login.microsoftonline.com/<Redacted ID>/oauth2/v2.0/token
    allowed_groups = <Redacted Group ID>
```

I can confirm that others users with membership of less than 200 groups does work. I added some logging statements in the [azure_oauth file in my forked repo](https://github.com/sivamu/grafana/blob/153ccbc49d3ce2bf9997f3347a3374befd254e1c/pkg/login/social/azuread_oauth.go#L62-L71) and for users that have a few groups, I see the following logs:

```
t=2021-12-01T19:50:30+0000 lvl=info msg="AzureAD OAuth: extracted groups" logger=azureoauth email=<email address> groups=[<space separated list of GUIDS>]
```

For myself, I see an empty list in the groups variable:
```
t=2021-12-01T19:48:18+0000 lvl=info msg="AzureAD OAuth: extracted groups" logger=azureoauth email=<my email address> groups=[]
```

**Comment by papagian** [comment]
@sivamu thank you for reporting and researching this.
According to [this](https://docs.microsoft.com/en-us/azure/active-directory/develop/access-tokens#payload-claims) I think a possible fix would be to modify [extractGrroups](https://github.com/grafana/grafana/blob/1008d463049d6d390fb536825942e44a6aff1a7e/pkg/login/social/azuread_oauth.go#L130) and if groups are empty to query endpoints under `_claim_sources` for fetching the user groups.

**Comment by kostrse** [comment]
Just for information, one more option is to use Graph API `me/checkMemberGroups` to check whether the user is a member of any given groups:
```
POST https://graph.microsoft.com/beta/me/checkMemberGroups
Content-type: application/json

{
  "groupIds": [
        "fee2c45b-915a-4a64b130f4eb9e75525e",
        "4fe90ae065a-478b9400e0a0e1cbd540"
  ]
}
```
PROs:
* Token itself doesn't need to contain any groups claims (in case if it's an issue)

CONs:
* Group IDs should be known (OIDs). Problematic of only names known. OIDs have to be obrained/extracted separately.

**Comment by Jguer** [comment]
This should be now be available with accompanying fixes in the latest version of 8.4, 8.5 and 9.0. 

https://github.com/grafana/grafana/pull/43470
https://github.com/grafana/grafana/pull/49506
